### DISCUSSION TITLE
move from BeforeRun, Run and AfterRun to Main

### DISCUSSION THREAD

**Discussion by AsertCreator** [CosmosOS/Cosmos#2625] [Ideas]
all of them can be replaced using `Main` method which can contain intialization, execution loop and shutdown if needed. `AfterRun` doesn't even make sense

**Comment by terminal-cs** [comment]
I prefer individual methods, it keeps things simple

**Comment by jspa2** [comment]
`AfterRun` is for shutdown/reboot tasks, like flushing databases and files

**Comment by quajak** [comment]
This would be a pretty big breaking change but if there is consensus that this simpler approach is better, I would be for changing it.

**Comment by MishaTy** [comment]
I agree with this change - and also all driver initialization should be moved to the user's kernel.

**Reply by terminal-cs** [reply]
> I agree with this change - and also all driver initialization should be moved to the user's kernel.

Good idea, now cosmos would require at least some knowledge and reading instead of lazy devs thinking they are new bill gates

**Reply by zarlo** [reply]
well it some what is its called in Start()

**Reply by ascpixi** [reply]
I'm also for making the end-user responsible for initializing the drivers they want to use - this would also enable IL2CPU to strip the code of the drivers that the end-user does not explicitly use.

**Reply by AsertCreator** [reply]
> I'm also for making the end-user responsible for initializing the drivers they want to use - this would also enable IL2CPU to strip the code of the drivers that the end-user does not explicitly use.

we need to create a simple driver model that will contain all drivers derived from some base class for all drivers from which developer can disable devices on need

**Comment by zarlo** [comment]
you can just use Start() if you want
